core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultPendingLockManager.java

```diff
@@ -0,0 +1,392 @@
+      throw new TimeoutException(format("Could not acquire lock on %s in behalf of transaction %s. Current owner %s.",
+                                        lockOwner.commonKey, thisGlobalTransaction,
+                                        lockOwner.cacheTransaction.getGlobalTransaction()));
+   }
+
+   private PendingTransaction waitForTransactionsToComplete(Collection<PendingTransaction> transactionsToCheck,
+                                                            long expectedEndTime) throws InterruptedException {
+      if (transactionsToCheck.isEmpty()) {
+         return null;
+      }
+      for (PendingTransaction tx : transactionsToCheck) {
+         long remaining;
+         if ((remaining = timeService.remainingTime(expectedEndTime, TimeUnit.MILLISECONDS)) > 0) {
+            if (!tx.cacheTransaction.waitForLockRelease(remaining)) {
+               return tx;
+            }
+         }
+      }
+      return null;
+   }
+
+   private Collection<PendingTransaction> getTransactionWithLockedKey(int transactionTopologyId,
+                                                                      Object key,
+                                                                      GlobalTransaction globalTransaction) {
+
+      Filter filter = transaction -> {
+         if (transaction.getTopologyId() < transactionTopologyId &&
+               !transaction.getGlobalTransaction().equals(globalTransaction) &&
+               transaction.containsLockOrBackupLock(key)) {
+            return key;
+         }
+         return null;
+      };
+
+      return filterAndCollectTransactions(filter);
+   }
+
+   private Collection<PendingTransaction> getTransactionWithAnyLockedKey(int transactionTopologyId,
+                                                                         Collection<Object> keys,
+                                                                         GlobalTransaction globalTransaction) {
+
+      Filter filter = transaction -> {
+         if (transaction.getTopologyId() < transactionTopologyId &&
+               !transaction.getGlobalTransaction().equals(globalTransaction)) {
+            return transaction.findAnyLockedOrBackupLocked(keys);
+         }
+         return null;
+      };
+
+      return filterAndCollectTransactions(filter);
+   }
+
+   private Collection<PendingTransaction> filterAndCollectTransactions(Filter filter) {
+      final Collection<? extends CacheTransaction> localTransactions = transactionTable.getLocalTransactions();
+      final Collection<? extends CacheTransaction> remoteTransactions = transactionTable.getRemoteTransactions();
+      final int totalSize = localTransactions.size() + remoteTransactions.size();
+      if (totalSize == 0) {
+         return Collections.emptyList();
+      }
+      final List<PendingTransaction> allTransactions = new ArrayList<>(totalSize);
+      final Consumer<CacheTransaction> consumer = transaction -> {
```

Comment:
I meant something like this: http://paste.fedoraproject.org/247332/65314214/

I think putting the filtering and collection in a single place actually makes it easier to read, you don't have to learn what a null return value means in the filter. Plus it allows us to remove the 'And' from the method name, which is always a bonus for me :)